File: leetcode/Binary_Tree_Inorder_Traversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ArrayList<Integer> inorderTraversal(TreeNode root) {
        ArrayList<Integer> ret = new ArrayList<Integer>();
        Stack<TreeNode> st = new Stack<TreeNode>();
        Stack<Boolean> ld = new Stack<Boolean>();
        if (root != null) {
            st.push(root);
            ld.push(false);
        }
        while (!st.empty()) {
            boolean f = ld.pop();
            if (!f) {
                ld.push(true);
                TreeNode t = st.peek();
                if (t.left != null) {
                    st.push(t.left);
                    ld.push(false);
                }
            } else {
                TreeNode t = st.pop();
                ret.add(t.val);
                if (t.right != null) {
                    st.push(t.right);
                    ld.push(false);
                }
            }
        }
        return ret;
    }
}

File: leetcode/Binary_Tree_Inorder_Traversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    void pushLeftSubtree(TreeNode *node, stack<TreeNode*> &printing) {
        while (node != NULL) {
            printing.push(node);
            node = node->left;
        }
    }
public:
    vector<int> inorderTraversal(TreeNode *root) {
        vector<int> inorder;
        stack<TreeNode*> printing;
        pushLeftSubtree(root, printing);
        while (!printing.empty()) {
            TreeNode *node = printing.top();
            printing.pop();
            inorder.push_back(node->val);
            pushLeftSubtree(node->right, printing);
        }
        return inorder;
    }
};
